### PR TITLE
Fix bug in timer (again)

### PR DESCRIPTION
**Pull request by samkumar**
This is the fix that Michael and I discussed on Friday. I also think I fixed a bug in the getAlarm() function in HplSam4lASTP, so please take a look at that too.

### CODE REVIEW

File: tos/chips/cortex/m4/sam4/l/timers/HalSam4lASTP.nc [hunk 1]
```diff
@@ -87,52 +85,43 @@ implementation
...
+            t1 = (t0 >> 1) + (dt >> 1) + (t0 & dt & 1);
+            t0 >>= 1;
+             
+            if (t0 > cv) {
+                cv |= (uint32_t) 0x80000000;
+            }
+            
+            if (t1 <= cv) {
+                // Fire the alarm right away
+                call ast.enable();
```
Comment (immesys) [inline tos/chips/cortex/m4/sam4/l/timers/HalSam4lASTP.nc:115]: you should add 

```
call ast.setCounterValue(cv);
```

above this line, to counter the effect of the tick-after-stop.

If you want it to be an unwrapped value, you should store the value you got from getCounterValue() before you wrapped it. Don't call getCounterValue() twice as it might not have the same value

File: tos/chips/cortex/m4/sam4/l/ast/HplSam4lASTP.nc [hunk 2]
```diff
@@ -203,6 +203,7 @@ implementation
 
 	async command uint32_t HplSam4lAST.getAlarm()
 	{
+	    while(AST->sr.bits.busy == 1);
```
Comment (immesys) [inline tos/chips/cortex/m4/sam4/l/ast/HplSam4lASTP.nc:206]: why did you add this?

File: tos/chips/cortex/m4/sam4/l/timers/HalSam4lASTP.nc [hunk 3]
```diff
@@ -87,52 +85,43 @@ implementation
...
+                signal Alarm.fired();
+            } else {
+                // Wait for t1 - cv ticks
                 call ast.setCounterValue(cv);
+                call ast.setAlarm(t1);
```
Comment (immesys) [inline tos/chips/cortex/m4/sam4/l/timers/HalSam4lASTP.nc:120]: The fix in getAlarm above is invalid. If you are concerned that setting the alarm is causing a race, the correct solution is to block until the set completes. I am not sure if the registers will sync while the clock is stopped, however. You should try and see if they will by modifying the setAlarm() function to include a while(busy) spin just before it returns. If the chip hangs indefinitely, then we know it doesn't progress while the AST is stopped. If it does not hang, then that will solve not only races with getAlarm() but also races with the AR==CV interrupt. As it stands it might be possible for CV to tick past AR before the register syncs

File: tos/chips/cortex/m4/sam4/l/ast/HplSam4lASTP.nc (hunk 2 above)
Comment (samkumar) [inline tos/chips/cortex/m4/sam4/l/ast/HplSam4lASTP.nc:206]: On Page 469 of the datasheet it says: "When the SR.BUSY bit is set writes to this register will be discarded and this register will read as zero." The phrase "this register" refers to the "ar0" register.

In setAlarm, the code correctly spin loops until busy is not set. In getAlarm, I didn't see this. Furthermore, when I call getAlarm, I got 0. So, I added this fix.

File: tos/chips/cortex/m4/sam4/l/timers/HalSam4lASTP.nc (hunk 3 above)
Comment (samkumar) [inline tos/chips/cortex/m4/sam4/l/timers/HalSam4lASTP.nc:120]: That actually wasn't my concern, but that is a good idea and I will try it out.

File: tos/chips/cortex/m4/sam4/l/timers/HalSam4lASTP.nc (hunk 1 above)
Comment (samkumar) [inline tos/chips/cortex/m4/sam4/l/timers/HalSam4lASTP.nc:115]: Isn't this already handled by the if statement in setAlarm?

**Comment by samkumar** [comment]
I will update my code and make another pull request shortly.
